rclcpp/topics/minimal_subscriber/member_function_with_type_adapter.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "rclcpp/type_adapter.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+using std::placeholders::_1;
+
+template<>
+struct rclcpp::TypeAdapter<std::string, std_msgs::msg::String>
+{
+  using is_specialized = std::true_type;
+  using custom_type = std::string;
+  using ros_message_type = std_msgs::msg::String;
+
+  static
+  void
+  convert_to_ros_message(
+    const custom_type & source,
+    ros_message_type & destination)
+  {
+    destination.data = source;
+  }
+
+  static
+  void
+  convert_to_custom(
+    const ros_message_type & source,
+    custom_type & destination)
+  {
+    destination = source.data;
+  }
+};
+
+
+class MinimalSubscriber : public rclcpp::Node
+{
+  using MyAdaptedType = rclcpp::TypeAdapter<std::string, std_msgs::msg::String>;
+
+public:
+  MinimalSubscriber()
+  : Node("minimal_subscriber")
+  {
+    subscription_ = this->create_subscription<MyAdaptedType>(
+      "topic", 10, std::bind(&MinimalSubscriber::topic_callback, this, _1));
+  }
+
+private:
+  void topic_callback(const std::shared_ptr<std::string> msg) const
+  {
+    RCLCPP_INFO(this->get_logger(), "I heard: '%s'", (*msg).c_str());
```

Comment:
If the previous comment is addressed this won't matter, but in the future (or if we don't do the other comment), this should be:

```suggestion
    RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg->c_str());
```